File: haskell/bin/2019/05.hs
import Advent (rawInput, solve)
import Advent.Intcode (Memory, parseMemory, run, vm)
import Advent.Parsing (parse)

readInput :: IO Memory
readInput = parse parseMemory <$> rawInput

part1 :: Memory -> Int
part1 = last . (`run` [1]) . vm

part2 :: Memory -> Int
part2 = last . (`run` [5]) . vm

main :: IO ()
main = solve readInput part1 part2

File: rust/src/bin/2019/05.rs
use advent_of_code::intcode::{IntcodeInterrupt, IntcodeRunner};
use std::time::Instant;

fn parse_input() -> Vec<isize> {
    include_str!("../../../../inputs/2019/05.txt").trim().split(",")
        .map(|s| s.parse().unwrap())
        .collect()
}

fn part1(program: Vec<isize>) -> isize {
    let mut runner = IntcodeRunner::for_program(program);
    runner.push_input(1);
    assert!(matches!(runner.run(), IntcodeInterrupt::Halt));
    runner.last_output().unwrap()
}

fn part2(program: Vec<isize>) -> isize {
    let mut runner = IntcodeRunner::for_program(program);
    runner.push_input(5);
    assert!(matches!(runner.run(), IntcodeInterrupt::Halt));
    runner.last_output().unwrap()
}

fn main() {
    let start_time = Instant::now();
    let program = parse_input();
    let part1_ans = part1(program.clone());
    let part2_ans = part2(program);
    let elapsed_time = start_time.elapsed();

    println!("Part 1: {}", part1_ans);
    println!("Part 2: {}", part2_ans);
    println!("Elapsed: {:?}", elapsed_time);
}
